### PR TITLE
gh-71052: Change Android's `sys.platform` from "linux" to "android"

### PR DESCRIPTION
**Pull request by mhsmith**
As discussed in [PEP 738](https://peps.python.org/pep-0738/#sys), although Android is based on Linux, it differs in enough significant ways that a separate name is justified.

Related changes:

* I've replaced all instances of `hasattr(sys, "getandroidapilevel")` with a `sys.platform` check. This is generally easier to read, and more consistent with how the other platforms are handled.

* I've checked all the places I could find that compare `sys.platform` to "linux", and added "android" where appropriate. These are mostly in the tests, but there are also a few in the stdlib.

* Issue: gh-71052

### CODE REVIEW

**Review by FFY00 — COMMENTED** [review]
Hum, I think this is a backwards incompatible change, which could result in downstream code breaking. While I think we should probably have a better way to identify Android platforms, I don't know exactly how I feel about this approach. I think at least we should get the input from people shipping and packaging Python on Android.

cc @freakboy3742 @misl6

**Comment by erlend-aasland** [comment]
I think you are right, Filipe. Sorry for being too eager in removing reviews from code owners; I'll take more care next time.

**Comment by mhsmith** [comment]
@freakboy3742 was actually the one who convinced me to go this way – his reasoning is on the PEP's Discourse thread [here](https://discuss.python.org/t/pep-738-adding-android-as-a-supported-platform/40975/6).

**Comment by freakboy3742** [comment]
> Hum, I think this is a backwards incompatible change, which could result in downstream code breaking. While I think we should probably have a better way to identify Android platforms, I don't know exactly how I feel about this approach. I think at least we should get the input from people shipping and packaging Python on Android.

As I said in the discussion thread on PEP 738 - I'm a *hard* +1 on `sys.platform == "android"`. 

Yes, this is technically backwards incompatible with the current state of Android support in the CPython source tree. However, I'd counter that:

1. CPython doesn't currently have *official* support for Android. It has defacto support by way of "not breaking the build", one additional API entry point (`sys.getandroidapilevel()`), and a handful of other conditional branches in code; but it's not a Tier 3 platform. To the extent that the ecosystem supports Android currently, that support is *very* ad hoc, and almost entirely performed by the people tagged in this thread (or at most one degree of separation from those people). There are no Android packages on PyPI.

2. The current state of detecting Android is the source of *so* many headaches in the rest of the ecosystem. Yes there is code that will "just run" because Android reports as Linux - but that code is almost always so trivial that it would likely also work on iOS - it's really testing for POSIX, not Linux. Any package that does anything even remotely complicated requires patching to run on Android - at the very least in the build system, and often in user-space code as well. The fact that `sys.platform` reports as Linux means we need to use workarounds [like this one](https://github.com/beeware/toga/blob/main/core/src/toga/platform.py#L29) everywhere that we do platform detection, because we can't just use a dictionary lookup. The only other platforms that cause this much headache are the ones that include the OS version in `sys.platform` - which is a pattern that (thankfully) appears to be slowly undergoing deprecation.

Repeating what [I said in the discuss thread](https://discuss.python.org/t/pep-738-adding-android-as-a-supported-platform/40975/6): Yes, technically, Android is a Linux. However, that is a distinction that isn't helpful in practice, because “Linux” isn’t just a kernel - it’s lots of other device expectations, filesystem expectations, userspace expectations, and more. 

The most practical manifestation of this: If Android is indeed "a Linux", why are any patches needed to CPython *at all*? Shouldn't Android "just work" as a Linux, or fall out of trivial configuration checks? Unless I'm mistaken, that there are less conditional branches in the test suite for *FreeBSD* support than are required for Android. If that isn't an indication that Android isn't "a Linux" in practice, I don't know what is.

File: Lib/test/test_os.py
Comment (erlend-aasland) [inline Lib/test/test_os.py]:
```suggestion
    SUPPORT_HEADERS_TRAILERS = (
        not sys.platform.startswith(("linux", "android", "solaris", "sunos")))
```

Comment (erlend-aasland) [reply Lib/test/test_os.py]: ... or perhaps:

```suggestion
    SUPPORT_HEADERS_TRAILERS = (
        not sys.platform.startswith(
            ("linux", "android", "solaris", "sunos")
        )
    )
```

File: Misc/NEWS.d/next/Build/2024-03-01-16-44-19.gh-issue-71052.Hs-9EP.rst
Comment (erlend-aasland) [inline Misc/NEWS.d/next/Build/2024-03-01-16-44-19.gh-issue-71052.Hs-9EP.rst]:
```suggestion
Change Android's :data:`sys.platform` from ``"linux"`` to ``"android"``.
```

**Review by erlend-aasland — APPROVED** [review]
Thanks for chiming in @freakboy3742.

Looks good to me. I had to look twice at the `test_sysconfig` change, though 😄

**Comment by mhsmith** [comment]
@erlend-aasland: I've applied your suggestions, so I think this should be ready to merge now.

**Review by FFY00 — CHANGES_REQUESTED** [review]
The PR looks good to me, the only thing missing is updating the documentation.

We should add Android to the "For other systems (...)" list in the `sys.platform` documentation (`Doc/library/sys.rst`), and add a `versionchanged` field noting this change.

After that, I think we are good to merge, thanks @mhsmith!

**Comment by mhsmith** [comment]
I have made the requested changes; please review again.

I've also cleaned up the `sys.platform` documentation as follows:

* Removed the reference to "append platform-specific components to `sys.path`". This has not been true since Python 3.7 (https://bugs.python.org/issue28046).
* Added iOS as specified in PEP 730.
* Sorted the table in alphabetical order.
* Moved the table to the top, since it now contains all supported platforms.
* Removed recommendation to use `startswith` on Linux and AIX. The last Python version to require this for Linux was 3.2, and the last for AIX was 3.7, both of which are now obsolete.

**Review by FFY00 — APPROVED** [review]
@mhsmith, sorry for the delay, the documentation changes look good to me, thank you!

**Pull request merged**
